easypy/threadtree.py

```diff
@@ -168,27 +177,77 @@ def iter_thread_frames():
         yield from sys._current_frames().items()
 
 
+class FrameWrapper:
+    """This wrapper adds the `f_thread_ident` attribute, which stores the thread
+    identifier of the current thread. All other attribute accesses are forwarded
+    to the wrapped frame object, making the wrapper behave like the original frame.
+    """
+
+    def __init__(self, frame, thread_ident):
+        self._frame = frame
+        self.f_thread_ident = thread_ident
+
+    def __getattr__(self, name):
+        """Delegate attribute access to the wrapped frame object."""
+        return getattr(self._frame, name)
+
+    def __setattr__(self, name, value):
+        """Allow setting attributes."""
+        if name == "_frame" or name == "f_thread_ident":
+            super().__setattr__(name, value)  # Handle internal attributes
+        else:
+            setattr(self._frame, name, value)  # Delegate to the wrapped frame
+
+    def __repr__(self):
+        """Provide a string representation including the class name and f_thread_ident."""
+        return (
+            f"<{self.__class__.__name__}(frame={repr(self._frame)}, "
+            f"f_thread_ident={self.f_thread_ident})>"
+        )
+
+    def __str__(self):
+        """Provide a readable string for the wrapper."""
+        return str(self._frame)
+
+    def __eq__(self, other):
+        """Equality comparison."""
+        if isinstance(other, self.__class__):
+            return (
+                self._frame == other._frame
+                and self.f_thread_ident == other.f_thread_ident
+            )
+        return self._frame == other
+
+    def __ne__(self, other):
+        """Inequality comparison."""
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        """Make the object hashable."""
+        return hash(self._frame)
+
+    def __dir__(self):
+        """List attributes of the wrapper and wrapped object."""
+        return dir(self._frame) + ["f_thread_ident"]
+
+
 def walk_frames(thread=None, *, across_threads=False):
     """
     Yields the stack frames of the current/specified thread.
 
     :param across_threads: yield frames of ancestor threads.
-    Note that the parent thread might be off to other things, and not actually in the frame that spawned the thread at the tip.
     """
 
     if not thread:
         thread = threading.current_thread()
 
-    frame = dict(iter_thread_frames()).get(thread.ident)
-
+    frame = sys._getframe(0)
     while frame:
-
-        yield frame
-
+        yield FrameWrapper(frame, thread.ident)
         frame = frame.f_back
         if not frame and across_threads and thread.parent:
+            frame = _FRAME_SNAPSHOTS_REGISTRY[thread.parent.ident][thread.ident]
```

Comment:
this is suprising... this implies that the frame object is still "alive", as well as all its parents?... are we effectively protecting these frames from the garbage-collector by putting the bottom frame in the registry? The frames also have pointers to other objects - we need to make sure this doesn't have too much impact on memory consumption.

I thought we would need to capture the traceback as strings, or something similar.

Reply:
I did not think about memory consumption, but you are right.
Tracebacks as strings are out of the picture because of how comet creates slash tracebacks. It must have an actual frames to do that.

Let me put some more thought into this.